From 8e086749cb648d5c26767949eea58d43eb3f67cc Mon Sep 17 00:00:00 2001
From: Prakash.Kumar <prakumar@habana.ai>
Date: Tue, 20 Jul 2021 06:28:54 +0300
Subject: [PATCH] [SW-42270] gaudi2: msix pba offset default value

This patch changes the value of msix pba default
value from 0x800 to 0x400. This is workaround for
HW errata which needs to be done before pcie link
is established.

Check HW JIRA H6-3241 for reference.

Change-Id: I832f7a589e4a43bf91863a796183ef72558fdd03
Signed-off-by: Prakash.Kumar <prakumar@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 66c72bc..a79b497 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -36,6 +36,7 @@
 #include "preboot/thermal.h"
 #include <arcboot_export.h>
 #include <pci.h>
+#include <gaudi2.h>
 
 #define PRODUCT_ID_STR_SIZE	0x8
 
@@ -98,6 +99,9 @@
 	init_prstn();
 
 	init_arc0();
+
+	/* This W/A is to be done before pcie link is established */
+	set_pci_msix_pba_offset_val();
 }
 
 void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 2e570b8..47f7d2a 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -36,6 +36,7 @@
 #include "preboot/thermal.h"
 
 #include "common.h"
+#include <gaudi2.h>
 
 
 #define ARC0_ICCM_BASE_ADDR		0x1000007FF7000000ull
@@ -143,6 +144,9 @@
 {
 	init_arc();
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
+
+	/* This W/A is to be done before pcie link is established */
+	set_pci_msix_pba_offset_val();
 }
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index ae1a7c2..462b8f3 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -30,6 +30,8 @@
 #include <gaudi2_blocks_linux_driver.h>
 #include "preboot/thermal.h"
 #include <string.h>
+#include <pcie_dbi_regs.h>
+#include "gaudi2.h"
 
 #include <common/pci_ids.h>
 
@@ -506,3 +508,10 @@
 	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
 			mmCPU_IF_PCIE_SPI_INTR_CLR);
 }
+
+#define MSIX_PBA_OFFSET_REG_VAL		0x400
+
+void set_pci_msix_pba_offset_val(void)
+{
+	regmap_writel(MSIX_PBA_OFFSET_REG_VAL, mmPCIE_DBI_MSIX_PBA_OFFSET_REG);
+}
diff --git a/arch/gaudi2/common/gaudi2.h b/arch/gaudi2/common/gaudi2.h
new file mode 100644
index 0000000..5d110a7
--- /dev/null
+++ b/arch/gaudi2/common/gaudi2.h
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+void set_pci_msix_pba_offset_val(void);
